### PR TITLE
Paypal Subscription

### PR DESCRIPTION
**Pull request by fattaholmanan**

### CODE REVIEW

File: lib/purchase/sub_scheduler.rb [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+module Purchase
+	class SubScheduler
+		include PaypalSubUtils
+
+		def self.tic_tac
+			orders = Array.new # [{id, expires_at}, ...]
+			
+			orders.concat apple_tic_tac
+			orders.concat android_tic_tac
+			orders.concat paypal_tic_tac
```
Comment (vahid-ashrafian) [inline lib/purchase/sub_scheduler.rb]: Doesn't PayPal have a call back on subscription?

File: lib/purchase/sub_scheduler.rb [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+			orders
+		end
+
+		def self.paypal_new_order agreement, s
+			deadline = PaypalSubUtils.agreement_expires_at agreement
+			auto_renew = agreement.state.eql? "Active"
+
+			# create new transaction if the new deadline is > now + margin_time
+			# We just want to make sure that new deadline is surely due to a re-newed subscription.
+			# Since it is obtained from google a miliseconds datetime, it could change during the time!
```
Comment (vahid-ashrafian) [inline lib/purchase/sub_scheduler.rb]: It's PayPal not google :))

File: lib/purchase/sub_scheduler.rb (hunk 1 above)
Comment (fattaholmanan) [inline lib/purchase/sub_scheduler.rb]: ?
Paypal has a `call_back` method which is being used when the user start the subscription. 
Here, we check new extended subscriptions. We then, in the logic module, create order based on the new extended subscriptions.

File: lib/purchase/sub_scheduler.rb (hunk 2 above)
Comment (fattaholmanan) [inline lib/purchase/sub_scheduler.rb]: Copy/Paste issue. Revised.

File: lib/purchase/sub_scheduler.rb [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+		def self.paypal_new_order agreement, s
+			deadline = PaypalSubUtils.agreement_expires_at agreement
+			auto_renew = agreement.state.eql? "Active"
+
+			# create new transaction if the new deadline is > now + margin_time
+			# We just want to make sure that new deadline is surely due to a re-newed subscription.
+			# To this end, we check wether it's beyond the margin_time from now. 
+			if deadline > (s.tran.expires_at + margin_time.days) # new receipt
+				s.expire
+				t = paypal_create_transaction s.tran, deadline
```
Comment (vahid-ashrafian) [inline lib/purchase/sub_scheduler.rb]: Is there any relational between PayPal's transaction_id and our own transaction_id, and if not is it possible to save PayPal's transaction_id?

**Pull request closed without merging**